physx/src/foundation.rs

```diff
@@ -281,3 +312,42 @@ unsafe impl AllocatorCallback for DefaultAllocator {
         unreachable!()
     }
 }
+
+unsafe extern "C" fn report_error_helper(
+    code: u32,
+    message: *const c_void,
+    file: *const c_void,
+    line: u32,
+    user_data: *const c_void,
+) {
+    let code = BitFlags::from_bits(code).unwrap_or_else(|_| {
+        debug_assert!(false, "bad error code {}", code);
+        Default::default()
+    });
+    let message =
+        String::from_utf8_lossy(unsafe { CStr::from_ptr(message.cast::<i8>()) }.to_bytes());
+    let file = String::from_utf8_lossy(unsafe { CStr::from_ptr(file.cast::<i8>()) }.to_bytes());
+    let ec = unsafe { &*user_data.cast::<Box<dyn ErrorCallback>>() };
+    ec.report_error(code, &message, &file, line);
+}
```

Comment:
Can we use the regular trampoline pattern instead? Not sure what this code does for us.

Reply:
what would the regular trampoline pattern look like?

Reply:
So if you look at e.g. the profiler trampoline, they don't use any helper functions like this. They replicate the C++ interface in a Rust trait, with an additional userdata at the end. Dealing with type conversions is left to the user of the library, because they might want to do different things, be on different platforms, etc.

Reply:
I figured it would be nice to have physx-rs provide safe apis and leave the pointer stuff to physx-sys as much as possible?

Reply:
Hmm. Yes. But I'm not sure if that should come at the cost of a double indirection. Wondering whether we could skip one layer of dynamic dispatch somehow...

Reply:
Maybe we could do this using a monomorphized shim instead, which would get rid of any dynamic calls here.

```
trait PhysxLoggingCallback {
    fn on_message<'a>(&self, msg: &'a str);
}


struct LoggingVTable {
    on_message: fn(state: *const c_void, msg: *const i8);
}

fn into_vtable<L: PhysxLoggingCallback>(callback: L) -> LoggingVtable {
    fn on_message_shim<L: PhysxLoggingCallback>(this: *const c_void, msg: *const i8) {
        let this = unsafe { &*this.cast::<L>() };
        let msg = OsStr::from_raw(msg).unwrap().as_str();
        this.on_message(msg);
    }
    
    LoggingVTable {
        on_message: on_message_shim<L>
    }
}
```

Reply:
(untested code, just off the top of my head for how this pattern generally looks)

Reply:
nice. yeah, that should avoid one indirection.
I'll try and do something like that.

Reply:
@jnises Sure! LMK if you want me to help out more actively :-)